indexserver: add debug endpoint for deleting repository shards

ggilmore opened the pull request:
This PR adds a new debug endpoint `/debug/delete?id=$REPO_ID`. When a user calls this URL, all the shards associated with this ID will be deleted (or tombstoned if they're in compound shards). 

Given the frequent [memory-map incidents](https://sourcegraph.slack.com/archives/C07KZF47K/p1668120346452419) that we had this this week, this PR should make it easier for users to recover when they're in this state. 

They should be able:

1. Grab a list of all IDs / repositories that are indexed on the instance:
	```bash
	wget -q -O - http://localhost:6072/debug/indexed                                                                                                                                              (base) 14:51:36
	ID              Name
	736075603       wut
	1309754292      whatever
	```
1. Extract some subset of repo ids via some method like `awk | tail/head`
	```bash
	wget -q -O - http://localhost:6072/debug/indexed | awk '{print $1}' | tail -n +2 | head -n 2                                                                                                  (base) 15:01:56
	736075603
	1309754292
	```
1. Pipe their chosen ids to this new debug route:
	```bash
	wget -q -O - http://localhost:6072/debug/indexed | awk '{print $1}' | tail -n +2 | head -n 2 | xargs -I {} -- wget -q -O -  "http://localhost:6072/debug/delete?id={}"
	```

This should also be useful for customers that want to try incremental indexing (and need a faster way to revert other than waiting for a new build). 

---

_(notes)_ 

- This PR still isn't final, I still need to add basic tests for `deleteShards`.

- I decided to put this logic entirely in `main.go`. This [logic is quite similar to the logic used in build/builder.go](https://sourcegraph.com/github.com/sourcegraph/zoekt/-/blob/build/builder.go?L759-777), but 
	- I didn't want to export this as a public function, I can't see anyone else needing to re-use this shard deletion logic (this debug endpoint should be the only user)
	- Builder has its own custom logic around shard deletion that's battle-tested (deleting shards only when it successfully finishes a new build). 
	
- I think the use of `build.Options` here isn't great. That object is so huge, and we ultimately only need the `indexDir`, `repoID`, and `repoName` out of it - everything else is irrelevant for this use case. I tried [experimenting](https://github.com/sourcegraph/zoekt/commit/aa80a5bc2c8fd3bb3cba7592bf0e965b24099415) with refactoring some of those interfaces, but I ended up just creating a bunch of [pass-through-methods](https://medium.com/@nathan.fooo/7-notes-different-layer-different-abstraction-fe0a64b04305) instead (from https://www.amazon.com/Philosophy-Software-Design-John-Ousterhout/dp/1732102201). What are your thoughts on this? Is it it worth the refactor? Can we create a smaller "options" object that doesn't have so many extraneous fields?

ggilmore committed:
indexserver: add debug endpoint for deleting repository shards

stefanhengl commented on cmd/zoekt-sourcegraph-indexserver/main.go:
```diff
@@ -757,6 +758,105 @@ func (s *Server) handleDebugList(w http.ResponseWriter, r *http.Request) {
+
+		metaFile := shard + ".meta"
+		if _, err := os.Stat(metaFile); err == nil {
+			err := os.Remove(metaFile)
```
We should remove the shard first, then the meta file. Otherwise you might end up serving wrong meta data, in the worst case leaking private or archived repos.

ggilmore replied:
That introduces a possibility that zoekt crashes in between removing a shard and its associated meta file. I guess we can add more logic in cleanup() to detect stranded meta files?

stefanhengl commented:
> I think the use of build.Options here isn't great. That object is so huge, and we ultimately only need the indexDir, repoID, and repoName out of it - everything else is irrelevant for this use case. I tried [experimenting](https://github.com/sourcegraph/zoekt/commit/aa80a5bc2c8fd3bb3cba7592bf0e965b24099415) with refactoring some of those interfaces, but I ended up just creating a bunch of [pass-through-methods](https://medium.com/@nathan.fooo/7-notes-different-layer-different-abstraction-fe0a64b04305) instead (from https://www.amazon.com/Philosophy-Software-Design-John-Ousterhout/dp/1732102201). What are your thoughts on this? Is it it worth the refactor? Can we create a smaller "options" object that doesn't have so many extraneous fields?

I would leave it as it. It is very "Zoekt" to pass around options structs. I believe you could do with IndexOptions but you need FindAllShards() which doesn't really have to implemented on top of build.Options and should maybe be a function instead.

However, I think at some point, Zoekt would profit from a unified concept of a shard IE operations (find(id), delete(id), ...) to abstract from compound shards, simple shards, and delta shards.

stefanhengl commented:
You might as well add your nice recipe "How to delete a repo" from the PR description to the documentation of the debug endpoint.

ggilmore committed:
add tests

ggilmore commented on cmd/zoekt-sourcegraph-indexserver/main_test.go:
```diff
@@ -124,6 +250,109 @@ func TestMain(m *testing.M) {
 	os.Exit(m.Run())
 }
 
+func createTestNormalShard(t *testing.T, indexDir string, r zoekt.Repository, numShards int, optFns ...func(options *build.Options)) []string {
```
mostly copied this from the builder package, should we create a `testutil` package?

ggilmore committed:
provide example shell pipeline in documentation

ggilmore commented:
@sourcegraph/search-core PTAL

gl-srgr commented on cmd/zoekt-sourcegraph-indexserver/main.go:
```diff
@@ -757,6 +760,112 @@ func (s *Server) handleDebugList(w http.ResponseWriter, r *http.Request) {
+
+			err := zoekt.SetTombstone(shard, options.RepositoryDescription.ID)
+			if err != nil {
+				return fmt.Errorf("setting tombstone in shard %q: %w", shard, err)
```
minor suggestion: Would it help to also log repo id that we were setting tombstone for?

stefanhengl commented on cmd/zoekt-sourcegraph-indexserver/main.go:
```diff
@@ -757,6 +760,112 @@ func (s *Server) handleDebugList(w http.ResponseWriter, r *http.Request) {
+		err = os.Remove(metaFile)
+		if err != nil {
+			return fmt.Errorf("deleting metadata file %q: %w", metaFile, err)
+		}
```
Is there a benefit of calling os.Stat before? os.Stat failing should be extremely rare.

```suggestion
		// remove the metadata file associated with the shard (if any)
		metaFile := shard + ".meta"

		err := os.Remove(metaFile)
		if err != nil {
		        if errors.Is(err, fs.ErrNotExist) {
				continue
			}

			return fmt.Errorf("deleting metadata file %q: %w", metaFile, err)
		}
```

stefanhengl reviewed (approved):
LGTM!

keegancsmith commented on cmd/zoekt-sourcegraph-indexserver/main.go:
```diff
@@ -757,6 +760,112 @@ func (s *Server) handleDebugList(w http.ResponseWriter, r *http.Request) {
+	repoID := uint32(id64)
+
+	s.queue.mu.Lock()
+	defer s.queue.mu.Unlock()
```
you are holding the queue lock while deleting. looks like you should only be holding it for the s.queue.get operation?

keegancsmith commented on cmd/zoekt-sourcegraph-indexserver/main.go:
```diff
@@ -757,6 +760,112 @@ func (s *Server) handleDebugList(w http.ResponseWriter, r *http.Request) {
+			}
+
+			return fmt.Errorf("'stat'ing metadata file %q: %w", metaFile, err)
+		}
```
use `zoekt.IndexFilePaths(shard)` to get a slice of all files associated with the shard.

keegancsmith reviewed (commented):
We already have very similiar logic to this in cleanup.go, which is run when we are told to stop indexing a repo. It feels like you should be able to reuse that? I took a look at cleanup and I believe it would be relatively straightforward. Additionally the function for this living in cleanup.go sounds good to me rather than growing main.go.

ggilmore committed:
move logic to cleanup.go

ggilmore committed:
wip - add logic to cleanup() to handle stranded metadata files

ggilmore committed:
switch deleteShards implementation to search across all shards in indexDir (instead of using buildOptions)

ggilmore committed:
wire up shardslog

ggilmore commented:
@sourcegraph/search-core PTAL

I have:

1) moved the logic from `main.go` into cleanup.go
2) changed the deletion logic to use zoekt.FilePaths instead of `stat`-ing directly
3) added a new sub-routine to cleanup() that ensures that we delete any metadata files that don't have corresponding shards (this is necessary after the change introduced in "2" since there is no longer an enforced order for how we delete files)
4) changed the logic for deleteShards to scan the entire indexDir instead of using `build.Options.FindAllShards()` (following the style of other functions in cleanup)
5) changed the debugHandler to take out a global indexDir lock 
6) wired up the shardLogger to the deleteShards implementation

ggilmore commented on cmd/zoekt-sourcegraph-indexserver/cleanup.go:
```diff
@@ -515,3 +549,68 @@ func removeTombstones(fn string) ([]*zoekt.Repository, error) {
+		// Is this repository inside a compound shard? If so, set a tombstone
+		// instead of deleting the shard outright.
+		if zoekt.ShardMergingEnabled() && maybeSetTombstone([]shard{s}, repoID) {
+			shardsLog(indexDir, "tomb", []shard{s})
```
Note: Looking at the implementation of shardslog, I'm unsure if it's threadsafe (what happens if there are two writers to the same file)? 

deleteShard's requirement that we hold the global index lock protects us against this, but I wonder if we should make this more explicit (e.g. have a dedicated mutex just for shardslog).

stefanhengl replied:
quoting: https://github.com/natefinch/lumberjack

> Lumberjack assumes that only one process is writing to the output files. Using the same lumberjack configuration from multiple processes on the same machine will result in improper behavior.

stefanhengl commented on cmd/zoekt-sourcegraph-indexserver/cleanup.go:
```diff
@@ -515,3 +549,68 @@ func removeTombstones(fn string) ([]*zoekt.Repository, error) {
+// isn't present in indexDir.
+//
+// Users must hold the global indexDir lock before calling deleteShards.
+func deleteShards(indexDir string, repoID uint32) error {
```
Quoting from cleanup:

```
	simple := shards[:0]
		for _, s := range shards {
			if shardMerging && maybeSetTombstone([]shard{s}, repo) {
				shardsLog(indexDir, "tombname", []shard{s})
			} else {
				simple = append(simple, s)
			}
		}

		if len(simple) == 0 {
			continue
		}

		removeAll(simple...)
```

Can't we just call that instead? I guess that is what @keegancsmith was refering to(?). We could factor it out into its own function and call it from your handler and from cleanup. WDYT?

ggilmore committed:
factor out cleanup and delete logic to shared function

ggilmore commented:
@stefanhengl PTAL at my latest commit which tries to factor out the logic. 

To be honest, I don't like the logic of this shared interface much at all (indexDir and shards are passed (and there is no contract that indexDir needs to contain those shards, indexDir is only passed for logging purposes, etc). LMK what you think.

keegancsmith commented on cmd/zoekt-sourcegraph-indexserver/cleanup.go:
```diff
@@ -172,6 +158,38 @@ func cleanup(indexDir string, repos []uint32, now time.Time, shardMerging bool)
 		}
 	}
 
+	// remove any Zoekt metadata files in the given dir that don't have an
+	// associated shard file
```
is this unrelated cleanup? IE maybe should be another PR?

ggilmore replied:
No, I added this functionality since we switched the implementation. Before, the implementation made sure to delete the metadata file _before_ its associated shard file. Since we switched to using zoekt.IndexFilePaths that order isn't guaranteed anymore (leaving open the possibility of a metadata file not having an associated shard if we delete the shard and then crash). I added this additional logic to cleanup so that we would have a background process that would reap those "stranded" files. 

I am happy to pull bit into a separate PR though.

keegancsmith commented on cmd/zoekt-sourcegraph-indexserver/cleanup.go:
```diff
@@ -292,11 +310,47 @@ func removeAll(shards ...shard) {
+	// - it's possible for zoekt to show inconsistent state when re-indexing the repository (zoekt incorrectly
+	//	 associates the data from shards 1 and 2 with the "new" shard 0 data (from a newer commit))
+	//
+	// Deleting shards in reverse sorted order (2 -> 1 -> 0) always ensures that we don't leave an inconsistent
+	// state behind even if we crash.
```
but if we do crash we will have partial state and be none the wiser anyways. So I think this is just extra complexity without making us any safer than before.

In fact I'd argue deleting 0 first is the best bet, since that is what we actually check in other parts. So if that is missing then we end up deleting the other ones. Then adding a check in cleanup to remove stranded indexes would make sense.

At the end of the day we are relying on the fact that os.Remove is super fast. Not ideal, and really we should introduce some other way of being atomic.

keegancsmith commented on cmd/zoekt-sourcegraph-indexserver/cleanup.go:
```diff
@@ -515,3 +569,26 @@ func removeTombstones(fn string) ([]*zoekt.Repository, error) {
+//
+// If one of the provided shards is a compound shard and the repository is contained within it,
+// the repository is tombstoned instead.
+func deleteOrTombstone(indexDir string, repoID uint32, shardMerging bool, shards ...shard) {
```
I see you mentioned you didn't like that we have to pass in indexDir and shardMerging to this func. Maybe instead it should be part of some helper struct that state. 

minor nit: I'd make `repoID` come after shardMerging param. IE the more "static" something is the closer to the front of the arg list. Just a habit from functional languages and currying.

ggilmore replied:
Got it...I'll play around with passing around some sort of helper to see if that's less awkward. Something like: 


```
type ServerConfig struct {
   shardMerging bool
   indexDir path

}

```

keegancsmith commented on cmd/zoekt-sourcegraph-indexserver/cleanup.go:
```diff
@@ -515,3 +569,26 @@ func removeTombstones(fn string) ([]*zoekt.Repository, error) {
+// If one of the provided shards is a compound shard and the repository is contained within it,
+// the repository is tombstoned instead.
+func deleteOrTombstone(indexDir string, repoID uint32, shardMerging bool, shards ...shard) {
+	var simple []shard
```
I see you got rid of the filtering hack, I guess just because now that it is a func it is safer that way?

ggilmore replied:
By filtering hack I assume you mean the `shardMap := getShards(indexDir)`  line? Yes, I decided to pass the list of shards associated with the ID directly since  `cleanup()` already has this information. If we're sharing the function implementation, gathering the entire `shardMap` seemed redundant.

keegancsmith commented on cmd/zoekt-sourcegraph-indexserver/cleanup.go:
```diff
@@ -292,11 +310,47 @@ func removeAll(shards ...shard) {
+	var sortedShards []shard
+	for _, s := range shards {
+		sortedShards = append(sortedShards, s)
+	}
```
can be a one liner
```suggestion
	sortedShards := append([]shard{}, shards...)
```

keegancsmith commented on cmd/zoekt-sourcegraph-indexserver/cleanup.go:
```diff
@@ -292,11 +310,47 @@ func removeAll(shards ...shard) {
 		paths, err := zoekt.IndexFilePaths(shard.Path)
 		if err != nil {
 			debug.Printf("failed to remove shard %s: %v", shard.Path, err)
+
```
nit
```suggestion
```

ggilmore committed:
Update cmd/zoekt-sourcegraph-indexserver/cleanup.go

Co-authored-by: Keegan Carruthers-Smith <keegan.csmith@gmail.com>

ggilmore committed:
Update cmd/zoekt-sourcegraph-indexserver/cleanup.go

Co-authored-by: Keegan Carruthers-Smith <keegan.csmith@gmail.com>

ggilmore committed:
handler: write successful message upon deleting a repository

keegancsmith commented:
What is the status of this PR? What can I do to help unblock/land?

jtibshirani commented:
@ggilmore do we still want to land this functionality? Or maybe the code has diverged too much and we should close?
